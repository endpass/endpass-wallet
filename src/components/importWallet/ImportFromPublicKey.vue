<template>
  <v-form>
    <v-input v-model="address"
             label="Address"
             id="address"
             name="address"
             v-validate="'required|address'"
             @input="handleInput"
             data-vv-as="private key"
             key="publicKeyUnique"
             aria-describedby="address"
             placeholder="0x...."
             required />

    <v-button className="is-primary is-medium"
              :loading="isCreating"
              @click.prevent="addWalletWithAddress">Import</v-button>
  </v-form>
</template>

<script>
import AddressWallet from '@/services/addressWallet';
import router from '@/router';
import { mapActions, mapMutations } from 'vuex';
import VForm from '@/components/ui/form/VForm.vue';
import VInput from '@/components/ui/form/VInput.vue';
import VButton from '@/components/ui/form/VButton.vue';

export default {
  name: 'import-from-public-key',
  data: () => ({
    isCreating: false,
    address: '',
  }),
  methods: {
    ...mapActions('accounts', ['addAccount']),
    ...mapMutations('accounts', ['setWallet']),
    handleInput() {
      this.errors.removeById('wrongAddress');
      this.$validator.validate();
    },
    async addWalletWithAddress() {
      this.isCreating = true;

      let wallet;

      try {
        wallet = this.createWalletWithAddress();
      } catch (e) {
        this.errors.add({
          field: 'address',
          msg: 'Address is invalid',
          id: 'wrongAddress',
        });
        console.error(e);
      }

      if (wallet) {
        try {
          await this.addAccount(wallet);
          router.push('/');
        } catch (e) {
          console.error(e);
        }
      }

      this.isCreating = false;
    },
    createWalletWithAddress() {
      return new AddressWallet(this.address);
    },
  },
  components: {
    VForm,
    VInput,
    VButton,
  },
};
</script>

<style lang="scss">
</style>
