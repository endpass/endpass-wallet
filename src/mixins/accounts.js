// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
