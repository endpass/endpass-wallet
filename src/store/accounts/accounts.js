import { userService } from '@/services';
import storage from '@/services/storage';
import web3 from 'web3';
import Bip39 from 'bip39';
import HDKey from 'ethereumjs-wallet/hdkey';
import { hdKeyMnemonic, kdfParams } from '@/config';
import EthWallet from 'ethereumjs-wallet';
import { SAVE_TOKENS } from '@/store/tokens/mutations-types';
import { Wallet, Address } from '@/class';
import { BigNumber } from 'bignumber.js';
import keystore from '@/utils/keystore';

export default {
  namespaced: true,
  state: {
    // TODO move to user store?
    email: null,
    // The encrypted xprv key for the user's hd wallet as a V3 keystore
    // object. All accounts are direct child accounts.
    hdKey: null,
    // The user's wallets, keyed by address
    wallets: {},
    // The currently selected wallet
    wallet: null,
    address: null,
    balance: null,
    price: null,
    // prettier-ignore
    availableCurrencies: ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
    settings: {
      fiatCurrency: 'USD',
    },
    otpSettings: {
      secret: null,
      status: null,
    },
  },
  getters: {
    isPublicAccount(state) {
      return (
        state.address instanceof Address && !(state.wallet instanceof Wallet)
      );
    },
    // TODO move to user store?
    isLoggedIn(state) {
      return !!state.email;
    },
    balance(state, getters, rootState, rootGetters) {
      if (state.balance === null) return null;

      const pendingBalance = rootGetters['transactions/pendingBalance'];
      const balanceWei = BigNumber(state.balance || '0')
        .minus(pendingBalance)
        .toFixed();

      return web3.utils.fromWei(balanceWei);
    },
    // Returns a decrypted HD Wallet
    hdWallet: state => password => {
      if (!state.hdKey) {
        return null;
      }
      return keystore.decryptHDWallet(password, state.hdKey);
    },
  },
  mutations: {
    setAddress(state, addressString) {
      state.address = new Address(addressString);
    },
    setWallet(state, wallet) {
      state.wallet = wallet;
    },
    addWallet(state, walletV3) {
      const wallet = new Wallet(walletV3);
      const { address } = walletV3;
      state.wallets = {
        ...state.wallets,
        [address]: wallet,
      };
    },
    // Saves the encrypted HD wallet key in V3 keystore format
    // Formerly addHdWallet
    setHdKey(state, key) {
      state.hdKey = key;
    },
    setBalance(state, balance) {
      state.balance = balance;
    },
    setPrice(state, price) {
      state.price = price;
    },
    setSettings(state, settings) {
      state.settings = JSON.parse(JSON.stringify(settings));
    },
    setEmail(state, email) {
      state.email = email;
    },
    setOtpSettings(state, otpSettings) {
      state.otpSettings = otpSettings;
    },
  },
  actions: {
    selectWallet({ commit, state, dispatch }, address) {
      commit('setWallet', state.wallets[address]);
      commit('setAddress', address);
      return dispatch(
        'tokens/subscribeOnTokensBalancesUpdates',
        {},
        { root: true },
      );
    },
    addWallet({ commit, dispatch, state }, json) {
      json.address = web3.utils.toChecksumAddress(json.address);
      commit('addWallet', json);

      return userService
        .setAccount(json.address, json)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    addWalletAndSelect({ dispatch }, json) {
      return dispatch('addWallet', json)
        .then(() =>
          dispatch('selectWallet', web3.utils.toChecksumAddress(json.address)),
        )
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    // Import wallet from json V3 keystore
    async addWalletWithV3({ commit, dispatch }, { json, password }) {
      try {
        const wallet = new Wallet(json);
        let privateKey = await wallet.getPrivateKey(password);
        return dispatch('addWalletWithPrivateKey', privateKey);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    async addWalletWithPrivateKey(
      { commit, dispatch },
      { privateKey, password },
    ) {
      try {
        const wallet = EthWallet.fromPrivateKey(Buffer.from(privateKey, 'hex'));
        const json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    generateWallet({ commit, dispatch, state, getters }, password) {
      if (!state.hdKey) {
        return;
      }
      try {
        let hdWallet = getters.hdWallet(password);
        let i = Object.keys(state.wallets).length;
        let wallet = hdWallet.deriveChild(i).getWallet();
        let json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    // Saves HD wallet's extended keys on the server
    saveHdWallet({ commit, dispatch, state }, json) {
      return userService.setAccount(json.address, json);
    },
    addHdWallet({ commit, dispatch }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);
      // Encrypt extended private key
      const json = keystore.encryptHDWallet(password, hdWallet);
      commit('setHdKey', json);

      // Save HD keys and generate the first child wallet
      return dispatch('saveHdWallet', json)
        .then(() => dispatch('generateWallet', password))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async addMultiHdWallet({ commit, dispatch, rootState }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);

      /* eslint-disable no-await-in-loop */
      for (let index = 0; index < 5; index++) {
        const wallet = hdWallet.deriveChild(index).getWallet();
        const walletV3 = wallet.toV3(Buffer.from(password), kdfParams);
        const { address } = walletV3;

        if (index !== 0) {
          dispatch('addWallet', walletV3);
        } else {
          dispatch('addWalletAndSelect', walletV3);
        }

        try {
          const balance = await rootState.web3.web3.eth.getBalance(address);

          if (balance === '0') {
            break;
          }
        } catch (e) {
          break;
        }
      }
      /* eslint-enable no-await-in-loop */
    },
    updateBalance({ commit, dispatch, state, rootState }) {
      if (state.address) {
        const address = state.address.getChecksumAddressString();

        return rootState.web3.web3.eth
          .getBalance(address)
          .then(balance => commit('setBalance', balance))
          .catch(e => dispatch('errors/emitError', e, { root: true }));
      }
    },
    updateSettings({ commit, dispatch }, settings) {
      commit('setSettings', settings);
      return storage
        .write('settings', settings)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    validatePassword({ state }, password) {
      return state.wallet.validatePassword(password);
    },
    login({ commit, dispatch }, email) {
      return userService.login(email);
    },
    logout({ commit, dispatch }) {
      commit('setEmail', null);
      return Promise.all([storage.clear(), userService.logout()])
        .then(() => window.location.reload())
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    loginViaOTP({}, { code, email }) {
      return userService.loginViaOTP(code, email);
    },
    getOtpSettings({ commit, dispatch }) {
      return userService
        .getOtpSettings()
        .then(otpSettings => commit('setOtpSettings', otpSettings))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    setOtpSettings({ commit, dispatch }, { secret, code }) {
      return userService
        .setOtpSettings(secret, code)
        .then(() => commit('setOtpSettings', { status: 'enabled' }))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    deleteOtpSettings({ commit, dispatch }, { code }) {
      return userService
        .deleteOtpSettings(code)
        .then(() => {
          commit('setOtpSettings', {});
          dispatch('getOtpSettings');
        })
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async init({ commit, dispatch }) {
      try {
        let [settings, email] = await Promise.all([
          userService.getSettings(),
          storage.read('email'),
        ]);

        commit(`tokens/${SAVE_TOKENS}`, settings.tokens || {}, { root: true });
        commit('setEmail', email);

        if (settings) {
          commit('setSettings', settings);
        }

        if (!email) {
          storage.disableRemote();
          return null;
        }

        // Fetch and save HD wallet
        let hdKey = await userService.getHDKey();
        if (hdKey) {
          commit('setHdKey', hdKey);
        }

        // Fetch and save regular accounts
        let accounts = await userService.getV3Accounts();
        if (accounts && accounts.length) {
          accounts.forEach(wallet => commit('addWallet', wallet));
          await dispatch('selectWallet', accounts[0].address);
        }
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
  },
};
