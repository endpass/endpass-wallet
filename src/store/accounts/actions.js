import { get, mapKeys, isEmpty } from 'lodash';
import {
  cryptoDataService,
  userService,
  localSettingsService,
} from '@/services';
import EthWallet from 'ethereumjs-wallet';
import { toChecksumAddress, bytesToHex } from 'web3-utils';
import { Wallet, NotificationError, web3 } from '@/class';
import { keystore } from '@endpass/utils';
import {
  CHANGE_INIT_STATUS,
  SET_ADDRESS,
  ADD_WALLET,
  REMOVE_WALLETS,
  SET_HD_KEY,
  SET_BALANCE,
  SET_HD_CACHE_BY_TYPE,
} from './mutations-types';

const WALLET_TYPES = Wallet.getTypes();
const WALLET_PROXY_TYPES = Wallet.getProxyTypes();

const selectWallet = async ({ commit, dispatch }, address) => {
  commit(SET_ADDRESS, toChecksumAddress(address));

  dispatch('updateBalance');
  dispatch('updateAccountSettings');
};

const addWallet = async ({ commit, dispatch }, json) => {
  try {
    const address = toChecksumAddress(json.address);
    const updatedJSON = { ...json, address };

    await userService.setAccount(address, updatedJSON);

    commit(ADD_WALLET, new Wallet(updatedJSON));
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const addPublicWallet = async (
  { dispatch },
  { address: rawAddress, info: extraInfo },
) => {
  try {
    const address = toChecksumAddress(rawAddress);
    const info = {
      type: WALLET_TYPES.PUBLIC,
      hidden: false,
      ...extraInfo,
      address,
    };

    await dispatch('addWalletAndSelect', { info, address });
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletAndSelect = async ({ dispatch }, json) => {
  try {
    await dispatch('addWallet', json);
    await dispatch('selectWallet', toChecksumAddress(json.address));
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

// Import wallet from json V3 keystore
const addWalletWithV3 = async (
  { dispatch },
  { json, jsonPassword, walletPassword },
) => {
  try {
    const wallet = new Wallet(json);
    const privateKey = await wallet.getPrivateKeyString(jsonPassword);

    return dispatch('addWalletWithPrivateKey', {
      privateKey,
      password: walletPassword,
    });
  } catch (e) {
    if (!(e instanceof NotificationError)) {
      throw e;
    }

    return dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletWithPrivateKey = async (
  { dispatch },
  { privateKey, password },
) => {
  try {
    const wallet = EthWallet.fromPrivateKey(
      Buffer.from(privateKey.replace(/^0x/, ''), 'hex'),
    );
    const v3KeyStore = keystore.encryptWallet(password, wallet, {
      kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
      n: ENV.VUE_APP_KDF_PARAMS_N,
    });

    return dispatch('addWalletAndSelect', v3KeyStore);
  } catch (e) {
    return dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletWithPublicKey = async ({ dispatch }, publicKeyOrAddress) => {
  // TODO convert public key to address, accept xPub key
  try {
    const address = toChecksumAddress(publicKeyOrAddress);
    const info = { type: WALLET_TYPES.PUBLIC };
    await dispatch('addWallet', { address, info });

    return dispatch('selectWallet', address);
  } catch (e) {
    return dispatch('errors/emitError', e, { root: true });
  }
};

const generateWallet = async ({ dispatch, state }, password) => {
  if (!state.hdKey) {
    throw new Error('hdKey doesn`t exist');
  }

  const decryptedHdWallet = await dispatch('decryptAccountHdWallet', password);
  const i = Object.keys(state.wallets).length;
  const wallet = decryptedHdWallet.deriveChild(i).getWallet();
  const v3KeyStore = keystore.encryptWallet(password, wallet, {
    kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
    n: ENV.VUE_APP_KDF_PARAMS_N,
  });

  await dispatch('addWalletAndSelect', v3KeyStore);
};

const commitWallet = async ({ commit }, { wallet }) => {
  if (keystore.isExtendedPublicKey(wallet.address)) {
    commit(SET_HD_KEY, wallet);
  } else {
    commit(ADD_WALLET, new Wallet(wallet));
  }
};

const saveWallet = async ({ dispatch }, { json, info = {} }) => {
  await userService.setAccount(json.address, { info, ...json });
  await dispatch('commitWallet', { wallet: json });
};

const addHdWallet = async ({ dispatch, getters }, { key, password }) => {
  try {
    const hdWallet = getters.getHdWalletBySeed(key);
    // Encrypt extended private key
    const v3KeyStore = keystore.encryptHDWallet(password, hdWallet, {
      kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
      n: ENV.VUE_APP_KDF_PARAMS_N,
    });
    const info = {
      address: v3KeyStore.address,
      type: WALLET_TYPES.HD_MAIN,
      hidden: false,
    };

    // Save HD keys and generate the first child wallet
    await dispatch('saveWallet', { json: v3KeyStore, info });
    await dispatch('generateWallet', password);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const addHdChildWallets = async (
  { dispatch, getters },
  { type, password, address, index },
) => {
  try {
    const v3KeyStore = getters.cachedHdV3KeyStoreByType(type);

    const hdWallet = keystore.decryptHDWallet(password, v3KeyStore);
    const wallet = hdWallet.deriveChild(index).getWallet();

    if (address !== wallet.getChecksumAddressString()) {
      throw new NotificationError({
        title: 'Add wallet',
        text:
          'Something goes wrong with during new wallet adding. Please try again.',
        type: 'is-danger',
      });
    }

    const v3KeyStoreChild = wallet.toV3(Buffer.from(password), {
      kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
      n: ENV.VUE_APP_KDF_PARAMS_N,
    });
    dispatch('addWalletAndSelect', v3KeyStoreChild);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const addHdPublicWallet = async (
  { commit, dispatch, getters },
  { key, password },
) => {
  try {
    const hdWallet = getters.getHdWalletBySeed(key);
    const v3KeyStore = keystore.encryptHDWallet(password, hdWallet, {
      kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
      n: ENV.VUE_APP_KDF_PARAMS_N,
    });
    const info = {
      address: v3KeyStore.address,
      type: WALLET_TYPES.HD_PUBLIC,
      hidden: false,
    };

    await userService.setAccount(v3KeyStore.address, {
      info,
      ...v3KeyStore,
    });

    commit(SET_HD_CACHE_BY_TYPE, {
      xpub: v3KeyStore.address,
      v3KeyStore,
      walletType: WALLET_TYPES.HD_PUBLIC,
    });
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const updateWallets = async ({ dispatch }, { wallets }) => {
  try {
    const { success } = await userService.updateAccounts(wallets);
    const promises = Object.values(wallets).map(wallet =>
      dispatch('commitWallet', { wallet }),
    );

    await Promise.all(promises);

    return success;
  } catch (error) {
    return dispatch('errors/emitError', error, { root: true });
  }
};

const getBalanceByAddress = async ({ rootGetters, rootState }, address) => {
  const { balance, tokens } = await cryptoDataService.getAccountBalance({
    network: rootGetters['web3/activeNetwork'],
    toSymbol: rootState.user.settings.fiatCurrency,
    address,
  });

  return {
    balance,
    tokens,
  };
};

const updateBalance = async ({ commit, dispatch, state }) => {
  if (!state.address) return;

  try {
    const { balance, tokens } = await dispatch(
      'getBalanceByAddress',
      state.address,
    );

    commit(SET_BALANCE, balance);
    dispatch(
      'tokens/setTokensInfoByAddress',
      {
        address: state.address,
        tokens,
      },
      {
        root: true,
      },
    );
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const validatePassword = async ({ state, getters }, password) => {
  if (getters.isPublicAccount || getters.isHardwareAccount) {
    const wallet = new Wallet(state.hdKey);

    return wallet.validatePassword(password);
  }

  return getters.wallet.validatePassword(password);
};

const setUserHdKey = async ({ commit, dispatch }) => {
  try {
    // Fetch and save HD wallet
    const hdKey = await userService.getHDKey();

    if (hdKey) {
      commit(SET_HD_KEY, hdKey);
      commit(SET_HD_CACHE_BY_TYPE, {
        xpub: hdKey.address,
        v3KeyStore: hdKey,
        walletType: WALLET_TYPES.HD_MAIN,
      });
    }
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const setUserWallets = async ({ commit, dispatch, rootState }) => {
  try {
    // Fetch and save regular accounts
    const accounts = await userService.getV3Accounts();

    if (isEmpty(accounts)) return;

    const localSettings = localSettingsService.load(rootState.user.email);
    const isAccountExist =
      localSettings &&
      localSettings.activeAccount &&
      accounts.find(({ address }) => address === localSettings.activeAccount);

    accounts.forEach(account => {
      commit(ADD_WALLET, new Wallet(account));
    });

    if (isAccountExist) {
      await dispatch('selectWallet', localSettings.activeAccount);
    } else {
      await dispatch('selectWallet', accounts[0].address);
    }
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const getNextWalletsFromHd = async (
  { state, dispatch, getters },
  { walletType, ...selectParams },
) => {
  const savedXpub = getters.cachedXpubByType(walletType);

  const params = {
    walletType,
    ...selectParams,
    xpub: savedXpub,
  };

  let proxyWallet;
  switch (walletType) {
    case WALLET_TYPES.TREZOR:
      proxyWallet = await Wallet.loadProxy(WALLET_PROXY_TYPES.TrezorProxy);
      break;
    case WALLET_TYPES.LEDGER:
      proxyWallet = await Wallet.loadProxy(WALLET_PROXY_TYPES.LedgerProxy);
      break;
    case WALLET_TYPES.HD_PUBLIC:
      proxyWallet = await Wallet.loadProxy(WALLET_PROXY_TYPES.HDProxy);
      break;
    case WALLET_TYPES.HD_MAIN:
      proxyWallet = await Wallet.loadProxy(WALLET_PROXY_TYPES.HDProxy);
      params.xpub = state.hdKey.address;
      break;
    default:
      throw new NotificationError({
        title: 'Access error',
        text:
          'An error occurred while getting access to hardware device. Please, try again.',
        type: 'is-danger',
      });
  }

  const result = await proxyWallet.getNextWallets(params);
  const { xpub, addresses } = result;

  if (savedXpub !== xpub) {
    await dispatch('saveToCache', { xpub, walletType });
  }

  return addresses;
};

const decryptAccountHdWallet = async ({ state }, password) => {
  if (!state.hdKey) {
    return null;
  }

  return keystore.decryptHDWallet(password, state.hdKey);
};

const decryptAccountWallets = async ({ state }, password) =>
  Object.values(state.wallets)
    .filter(item => !item.isPublic && !item.isHardware)
    .map(item => keystore.decryptWallet(password, item.v3));

const encryptHdWallet = async (ctx, { password, hdWallet }) => {
  if (!hdWallet) return null;

  return keystore.encryptHDWallet(password, hdWallet, {
    kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
    n: ENV.VUE_APP_KDF_PARAMS_N,
  });
};

const encryptWallets = async (ctx, { password, wallets = [] }) =>
  wallets.map(item =>
    keystore.encryptWallet(password, item, {
      kdf: ENV.VUE_APP_KDF_PARAMS_KDF,
      n: ENV.VUE_APP_KDF_PARAMS_N,
    }),
  );

const reencryptAllAccountWallets = async (
  { dispatch },
  { password, newPassword },
) => {
  const [decryptedHdWallet, decryptedWallets] = await Promise.all([
    dispatch('decryptAccountHdWallet', password),
    dispatch('decryptAccountWallets', password),
  ]);
  const encryptedHdWallet = await dispatch('encryptHdWallet', {
    hdWallet: decryptedHdWallet,
    password: newPassword,
  });
  const encryptedWallets = await dispatch('encryptWallets', {
    wallets: decryptedWallets,
    password: newPassword,
  });

  return {
    hdWallet: encryptedHdWallet,
    wallets: encryptedWallets,
  };
};

const updateWalletsWithNewPassword = async (
  { dispatch },
  { password, newPassword },
) => {
  const { hdWallet, wallets } = await dispatch('reencryptAllAccountWallets', {
    password,
    newPassword,
  });
  const walletsToUpdate = mapKeys(wallets, 'address');

  if (hdWallet) {
    Object.assign(walletsToUpdate, {
      [hdWallet.address]: hdWallet,
    });
  }

  if (isEmpty(walletsToUpdate)) {
    return null;
  }

  const res = await dispatch('updateWallets', {
    wallets: walletsToUpdate,
  });

  return res;
};

const recoverWalletsPassword = async (
  { state, dispatch, commit, getters },
  { seedPhrase, password },
) => {
  try {
    if (!state.hdKey) {
      throw new NotificationError({
        title: 'Error recovering wallet password',
        text: 'Main HD wallet not found.',
        type: 'is-danger',
      });
    }

    const hdWallet = getters.getHdWalletBySeed(seedPhrase);
    const hdWalletAddress = hdWallet.publicExtendedKey();

    if (state.hdKey.address !== hdWalletAddress) {
      throw new NotificationError({
        title: 'Error recovering wallet password',
        text: 'Incorrect seed phrase.',
        type: 'is-danger',
      });
    }

    const passswordRecoveryIdentifier = await userService.getPasswortRecoveryIdentifier();
    const wallet = hdWallet.deriveChild(0).getWallet();
    const privateKey = bytesToHex(wallet.getPrivateKey());
    const { signature } = await web3.eth.accounts.sign(
      passswordRecoveryIdentifier,
      privateKey,
    );
    const encryptedHdWallet = await dispatch('encryptHdWallet', {
      hdWallet,
      password,
    });
    const encryptedWallets = await dispatch('encryptWallets', {
      wallets: [wallet],
      password,
    });
    const { success } = await userService.recoverWalletsPassword({
      signature,
      main: {
        address: hdWalletAddress,
        keystore: encryptedHdWallet,
      },
      standart: {
        address: toChecksumAddress(bytesToHex(wallet.getAddress())),
        keystore: encryptedWallets[0],
      },
    });

    if (success) {
      const successMessage = new NotificationError({
        title: 'The password is successfully recovered',
        text:
          'All keystore wallets have been deleted. They can be restored from a private key or seed phrase on the import page.',
        type: 'is-success',
      });

      commit(REMOVE_WALLETS);
      await Promise.all([dispatch('setUserHdKey'), dispatch('setUserWallets')]);
      await dispatch('errors/emitError', successMessage, { root: true });
    }
  } catch (error) {
    await dispatch('errors/emitError', error, { root: true });
  }
};

const saveToCache = async ({ commit }, { xpub, walletType }) => {
  const info = { type: walletType };

  commit(SET_HD_CACHE_BY_TYPE, { xpub, walletType });

  await userService.setAccount(xpub, { info });
};

const updateAccountSettings = async ({
  state,
  rootGetters,
  rootState,
  dispatch,
}) => {
  const lastActiveAccount = rootGetters['user/lastActiveAccount'];
  const { address } = state;
  const wallet = get(state, `wallets[${address}]`, {});
  const isNotPublicAccount = wallet.isPublic === false;

  if (address !== lastActiveAccount && isNotPublicAccount) {
    await dispatch(
      'user/updateSettings',
      {
        lastActiveAccount: address,
      },
      { root: true },
    );
  }

  localSettingsService.save(rootState.user.email, {
    activeAccount: address,
  });
};

const backupSeed = async ({ getters, dispatch }, { seed, password }) => {
  try {
    const encryptedSeed = await getters.wallet.encryptMessageWithPublicKey(
      seed,
      password,
    );

    await userService.backupSeed(encryptedSeed);
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const recoverSeed = async ({ getters }, password) => {
  const encryptedSeed = await userService.recoverSeed();
  const recoveredSeed = await getters.wallet.decryptMessageWithPrivateKey(
    encryptedSeed,
    password,
  );

  return recoveredSeed;
};

const init = async ({ commit, dispatch }) => {
  try {
    await Promise.all([dispatch('setUserHdKey'), dispatch('setUserWallets')]);
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }

  commit(CHANGE_INIT_STATUS, true);
};

export default {
  selectWallet,
  addWallet,
  addWalletAndSelect,
  addWalletWithV3,
  addWalletWithPrivateKey,
  addWalletWithPublicKey,
  addPublicWallet,
  commitWallet,
  saveWallet,
  generateWallet,
  setUserHdKey,
  setUserWallets,
  addHdWallet,
  addHdPublicWallet,
  addHdChildWallets,
  updateWallets,
  updateBalance,
  getBalanceByAddress,
  validatePassword,
  getNextWalletsFromHd,
  saveToCache,
  decryptAccountHdWallet,
  decryptAccountWallets,
  encryptHdWallet,
  encryptWallets,
  reencryptAllAccountWallets,
  updateWalletsWithNewPassword,
  recoverWalletsPassword,
  updateAccountSettings,
  backupSeed,
  recoverSeed,
  init,
};
