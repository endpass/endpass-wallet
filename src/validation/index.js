import Vue from 'vue';
import { default as VeeValidate, Validator } from 'vee-validate';
import address from './address';
import privateKey from './privateKey';
import publicKey from './publicKey';
import seedPhrase from './seedPhrase';
import ensOrAddress from './ensOrAddress';
import hex from './hex';

Validator.extend('address', address);
Validator.extend('ensOrAddress', ensOrAddress, {
  immediate: false,
});
Validator.extend('private_key', privateKey);
Validator.extend('public_key', publicKey);
Validator.extend('seed_phrase', seedPhrase);
Validator.extend('hex', hex);

Vue.use(VeeValidate, { inject: false });
