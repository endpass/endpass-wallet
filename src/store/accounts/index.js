import actions from './actions';
import getters from './getters';
import mutations from './mutations';

/**
 * @namespace
 * @property {Boolean} isInited Status of initializing accounts module
 * @property {Object} hdKey The encrypted xprv key for the user's hd wallet as a V3 keystore object.
 *  All accounts are direct child accounts.
 * @property {Object} hadrwareXpub The user's hardware xpub, keyed by wallet type
 * @property {Object} wallets The user's wallets, mapped by address
 * @property {String} address Current wallet address
 * @property {String} balance Currect account balance
 */
const state = {
  isInited: false,
  hardwareXpub: {},
  wallets: {},
  hdKey: null,
  address: null,
  balance: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
