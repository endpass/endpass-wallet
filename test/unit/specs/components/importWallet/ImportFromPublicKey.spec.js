import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import VueRouter from 'vue-router';
import UIComponents from '@endpass/ui';
import validation from '@/validation';

import { address } from 'fixtures/accounts';

import ImportFromPublicKey from '@/components/importWallet/ImportFromPublicKey';

const localVue = createLocalVue();

localVue.use(validation);
localVue.use(Vuex);
localVue.use(VeeValidate);
localVue.use(VueRouter);
localVue.use(UIComponents);

describe('ImportFromPublicKey', () => {
  let wrapper;
  let actions;
  let router;
  beforeEach(() => {
    actions = {
      addWalletWithPublicKey: jest.fn(),
    };
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions,
        },
      },
    };
    router = new VueRouter();
    const store = new Vuex.Store(storeOptions);
    wrapper = shallowMount(ImportFromPublicKey, {
      localVue,
      store,
      router,
      sync: false,
    });
  });
  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('ImportFromPublicKey');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('addWallet', () => {
      it('should call vuex addWalletWithPublicKey with correct arguments', () => {
        wrapper.setData({ address });
        wrapper.vm.submitAddWallet();

        expect(actions.addWalletWithPublicKey).toBeCalledWith(
          expect.any(Object),
          address,
          undefined,
        );
      });

      it('should redirect to root after successful wallet creation', async () => {
        expect.assertions(2);

        router.push('/kek');
        expect(router.currentRoute.fullPath).toBe('/kek');

        await wrapper.vm.submitAddWallet();
        expect(router.currentRoute.fullPath).toBe('/');
      });

      it('should add error to field if failed to create wallet', () => {
        expect.assertions(1);

        actions.addWalletWithPublicKey.mockImplementationOnce(() => {
          throw new Error();
        });
        wrapper.vm.submitAddWallet();

        expect(wrapper.vm.errors.items[0]).toMatchObject({
          field: 'address',
          id: 'wrongAddress',
          msg: 'Address is invalid',
        });
      });
    });

    describe('handleInput', () => {
      it('should clear error with wrongAddress id', () => {
        wrapper.vm.errors.add({
          field: 'address',
          msg: 'Address is invalid',
          id: 'wrongAddress',
        });
        expect(wrapper.vm.errors.has('address')).toBe(true);

        wrapper.vm.handleInput();
        expect(wrapper.vm.errors.has('address')).toBe(false);
      });
    });
  });
});
