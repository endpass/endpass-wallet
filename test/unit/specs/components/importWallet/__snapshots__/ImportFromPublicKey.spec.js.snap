// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportFromPublicKey render should render initial state of the component 1`] = `
<v-form
  data-test="import-public-form"
>
  <v-input
    aria-describedby="address"
    aria-invalid="false"
    aria-required="true"
    data-test="input-public-key"
    data-vv-as="public key"
    data-vv-name="address"
    id="address"
    label="Address"
    name="address"
    placeholder="0x...."
    required=""
  />
   
  <v-button
    class-name="is-primary is-cta"
    data-test="submit-import"
    disabled="true"
  >
    
    Import
  
  </v-button>
</v-form>
`;
